### DISCUSSION TITLE
How to model dependent atoms that need a cleanup/destructor/teardown step?

### DISCUSSION THREAD

**Discussion by folz** [pmndrs/jotai#2025] [Q&A] [answered]
Hi, I'm trying to model a [REGL](https://github.com/regl-project/regl) instance in Jotai. REGL depends on the existence of a canvas element first, so I have:

```ts
// This gets set to a `<canvas>` when a related view is mounted.
const canvas$ = atom<HTMLCanvasElement | null>(null);

const regl$ = atom(get => {
  const canvas = get(canvas$);

  if (!canvas) return undefined;
  
  return REGL({ canvas, ... });
});
```

My issue is that I need to call `get(regl$).destroy()` whenever the old regl instance is replaced with a new one, or I get WebGL memory leaks.

Initially I tried using onMount, thinking I would do nothing in the `onMount` and call this in the returned `onUnmount`. However, dependent atoms do not allow `onMount`, and anyway `onUnmount` does not receive any arguments, so it cannot reference the instance.

I think there may be a way to do this with making `canvas$` a writable atom, and having it call `get(regl$).destroy()` before writing canvas. However, I feel like the canvas atom should not need to be aware of the fact that "regl needs a destructor", since it does not otherwise need to be aware of regl.

How should I think about this dependent cleanup/destructor pattern using Jotai? Any help or pointers are appreciated. Thanks for a great library!

**Comment by dai-shi** [comment]
Well, you are hitting a challenging point of this library.
So, we basically assume that the `read` function is declarative and pure. It's the same as how React render function works. You would have similar difficulty with `useState` too.
There would be many cases, but the reason it needs `.destroy()` implies that it does some side effects on creation, which means impure or not garbage collectable...

Enough about the theory, what would we do? Is it the first time we face this issue? Of course not.
We had the similar problem with all those subscription based integrations, such as jotai-tanstack-query, and jotai-urql.
We ended up with a hack. The hack is in `atomWithObservable` and all those integrations use it.
There's a risk of memory leaks. (To mitigate it, there's `unstable_timeout` option, but not widely used.)

So, there are no canonical solutions, only hacks.
If you are sure it's mounted, and should only destroy on unmount, it's rather easy.
```js
const regl$ = atom(get => {
  const canvas = get(canvas$);
  if (!canvas) return undefined;
  return REGL({ canvas, ... });
}, (get, set) => {
  return () => get(regl$)?.destory();
});
regl$.onMount = (init) => {
  const destroy = init();
  return destroy;
};
```
(It doesn't cover all cases.)

**Comment by dai-shi** [comment, marked as answer]
There's a canonical solution. We shouldn't use `read` with impure code. We should use `write`.

```js
const canvasAndRegl = atom(null);

const canvas$ = atom(
  (get) => get(canvasAndRegl)?.[0],
  (get, set, canvas) => {
    // here we can destroy old instance
    const regl = REGL({ canvas, ... });
    set(canvasAndRegl, [canvas, regl]);
  }
);

const regl$ = atom((get) => get(canvasAndRegl)?.[1]);
```

**Comment by folz** [comment]
Thanks for the response. Agree it's challenging to do with useState. My pre-jotai approach was

```tsx
const regl = useMemo(..., [canvas]);
useEffect(() => () => regl.destroy(), [regl])
```

This seems like it might relate to discussion in https://github.com/pmndrs/jotai/issues/211 too, about atom effects. I appreciate the examples you left here and in that thread, and I think I have a better sense for how to approach this problem.
